nixos/modules/programs/direnv.nix

```diff
@@ -0,0 +1,136 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}: let
+  cfg = config.programs.direnv;
+in {
+
+  options.programs.direnv = {
+
+    enable = lib.mkOption {
+
+      type = lib.types.bool;
+
+      default = false;
+
+      example = true;
+
+      description = ''
+        Whether to enable direnv integration. Takes care of both installation and
+        setting up the sourcing of the shell. Additionally enables nix-direnv
+        integration. Note that you need to logout and login for this change to apply.
+      '';
+    };
+
+    direnvrcExtra = lib.mkOption {
+
+      type = lib.types.lines;
+
+      default = "";
+
+      example = ''
+        export FOO="foo"
+        echo "loaded direnv!"
+      '';
+
+      description = ''
+        Extra lines to append to the sourced direnvrc
+      '';
+    };
+
+    silent = lib.mkOption {
+
+      type = lib.types.bool;
+
+      default = false;
+
+      example = true;
+
+      description = ''
+        Whether to hide direnv logging
+      '';
+    };
+
+    persistDerivations = lib.mkOption {
+
+      type = lib.types.bool;
+
+      default = true;
+
+      example = false;
+
+      description = ''
+        Whether to set keep-derivations and keep-outputs to true
+        to prevent shells from getting garbage collected
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    programs = {
+
+      zsh.interactiveShellInit = ''
+        eval "$(direnv hook zsh)"
+      '';
+
+      bash.interactiveShellInit = ''
+        eval "$(direnv hook bash)"
+      '';
+
+      fish.interactiveShellInit = ''
+        direnv hook fish | source;
+      '';
+
+    };
+
+    nix.settings = lib.mkIf cfg.persistDerivations {
+      keep-outputs = true;
+      keep-derivations = true;
+    };
+
+    environment = {
+
+      systemPackages = [pkgs.direnv];
+
+      variables = {
+        DIRENV_CONFIG = "/etc/direnv";
+
+        DIRENV_LOG_FORMAT = lib.mkIf cfg.silent "";
+
+      };
+
+      etc = {
+        "direnv/direnvrc".text = ''
+          #Load nix-direnv
+          source ${pkgs.nix-direnv}/share/nix-direnv/direnvrc
+
+          #Load direnvrcExtra
+          ${cfg.direnvrcExtra}
+
+          #Load ~/.direnvrc if present
```

Comment:
This is a bit confusing. Maybe something like this? 
```diff
- #Load ~/.direnvrc if present
+ #Load user-configuration if present (~/.direnvrc or ~/.config/direnv/direnvrc)
```

Reply:
Done!